src/hp/derp/http/server.rs

```diff
@@ -1,50 +1,404 @@
+            Some(f) => f,
+            None => Box::new(move |mut res: ResponseBuilder| {
+                for (k, v) in h.iter() {
+                    res = res.header(*k, *v);
+                }
+                let r = res
+                    .status(StatusCode::NOT_FOUND)
+                    .body(b"Not Found"[..].into())
+                    .unwrap();
+                HyperResult::Ok(r)
+            }),
+        };
+
+        let service = DerpService::new(
+            self.handlers,
+            derp_handler,
+            self.derp_endpoint,
+            not_found_fn,
+            self.headers,
+        );
+
+        let server_state = ServerState {
+            addr: self.addr,
+            tls_config: self.tls_config,
+            server: derp_server,
+            service,
+        };
+
+        server_state.serve().await
+    }
+}
+
+impl std::fmt::Debug for ServerBuilder {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let hyper_fn =
+            "Some(Box<dyn Fn(ResponseBuilder) -> HyperResult<Response<Body>> + Send + Sync + 'static)";
+        let derp_override = if let Some(_) = self.derp_override {
+            hyper_fn
+        } else {
+            "None"
+        };
+
+        let not_found_fn = if let Some(_) = self.not_found_fn {
+            hyper_fn
+        } else {
+            "None"
+        };
 
-    // send to the derp server
-    conn_handler.accept(reader, writer).await
+        write!(f, "ServerBuilder {{ secret_key: {:?}, addr: {:?}, mesh_key: {:?}, tls_config: {:?}, handlers: {:?}, derp_endpoint: {:?}, derp_override: {derp_override}, headers: {:?}, not_found_fn: {not_found_fn}  }}", self.secret_key, self.addr, self.mesh_key, self.tls_config, self.handlers, self.derp_endpoint, self.headers)
+    }
+}
+
+#[derive(Debug)]
+pub struct ServerState {
+    addr: SocketAddr,
+    tls_config: Option<TlsConfig>,
+    server: Option<crate::hp::derp::server::Server<HttpClient>>,
+    service: DerpService,
 }
 
-impl<P> hyper::service::Service<Request<Body>>
-    for ClientConnHandler<OwnedReadHalf, OwnedWriteHalf, P>
+impl ServerState {
+    async fn serve(self) -> Result<Server> {
+        match &self.tls_config {
+            Some(_) => self.https_serve().await,
+            None => self.http_serve().await,
+        }
+    }
+
+    /// Binds a TCP listener on `addr` and handles content using HTTP.
+    /// Returns the local `SocketAddr` on which the server is listening.
+    async fn http_serve(self) -> Result<Server> {
+        let http_listener = TcpListener::bind(&self.addr)
+            .await
+            .context("failed to bind https")?;
+        let addr = http_listener.local_addr()?;
+        let cancel = CancellationToken::new();
+        info!("[HTTP] derp: serving on {addr}");
+        let c = cancel.clone();
+        let task = tokio::task::spawn(async move {
+            debug!("about to loop");
+            loop {
+                match http_listener.accept().await {
```

Comment:
Does this loop terminate when the cancel token in cancelled?

Reply:
I was handling shutdown in a strange way and this comment nudged me to do better, so thank you! 

Now, when we call `server.shutdown()`:
- We use the cancel token to close the loop in the task that listens for incoming connections (called this `http_server_task`)
- Each incoming connection caused a new task to be spawned on a `JoinSet`. That `JoinSet` is cleaned up in the `http_server_task` task when the loop is closed. 
- Then we wait for the `http_server_task` to finish, and log any errors that occur when attempting to shut it down.

I think this is much cleaner, however, it still feels a bit roundabout. Suggestions welcome.

Reply:
seems very reasonable and easy to follow.  i wouldn't say roundabout